### PR TITLE
[9.x] Add new and improved AblyBroadcaster

### PR DESCRIPTION
**Pull request by owenpearson**
**Note: this PR is one of four parallel pull requests to add first class support for Ably to Laravel. Please find the other PRs here:**
- laravel/docs#8120
- laravel/laravel#5969
- laravel/echo#351

Background: the current Ably broadcaster uses the Pusher JavaScript SDK and the documentation instructs users to use Ably with the Pusher compatibility mode enabled. We’ve received a lot of feedback that various features of the Ably Laravel integration don’t work as expected, often caused by the fact that there’s a dependency on the Pusher JavaScript SDK which we have no control over. This collection of PRs adds a new broadcaster which uses the ably-js SDK which means that we can ensure that the Ably broadcaster will remain stable. We are more than happy to take feedback and answer any questions about the work presented here :)

- Adds a new AblyBroadcaster (enabled by the `ABLY_NATIVE_CLIENT ` environment variable) class compatible with the AblyConnector from laravel/echo#351 and associated tests.
Makes it possible to use [Ably capabilities](https://ably.com/docs/core-features/authentication#capabilities-explained) when authenticating clients to specify which operations they can perform on specified channels.
- See laravel/docs#8120 for usage instructions.

### CODE REVIEW

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 1]
```diff
@@ -22,108 +23,161 @@ class AblyBroadcaster extends Broadcaster
...
+     */
+    private $defaultChannelClaims = [
+        'public:*' => ['subscribe', 'history', 'channel-metadata'],
+    ];
+
     /**
      * Create a new broadcaster instance.
      *
-     * @param  \Ably\AblyRest  $ably
+     * @param $ably \Ably\AblyRest
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:47]:
```suggestion
     * @param  \Ably\AblyRest  $ably
     * @param  array $config
```
First variable type and then the variable name
The second parameter is missing (probably an array)

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 2]
```diff
@@ -22,108 +23,161 @@ class AblyBroadcaster extends Broadcaster
...
+    private static $serverTimeDiff = null;
+
     /**
-     * Authenticate the incoming request for a given channel.
-     *
-     * @param  \Illuminate\Http\Request  $request
-     * @return mixed
-     *
-     * @throws \Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException
+     * @param $time int
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:67]:
```suggestion
     * @param  int  $time
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 3]
```diff
@@ -22,108 +23,161 @@ class AblyBroadcaster extends Broadcaster
...
+    protected function getPrivateToken()
+    {
+        return Str::after($this->ably->options->key, ':');
+    }
 
-        $user = $this->retrieveUser($request, $channelName);
+    /**
+     * Authenticate the incoming request for a given channel.
+     *
+     * @param $request \Illuminate\Http\Request
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:110]:
```suggestion
     * @param  \Illuminate\Http\Request  $request
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 4]
```diff
@@ -22,108 +23,161 @@ class AblyBroadcaster extends Broadcaster
...
     /**
-     * Generate the signature needed for Ably authentication headers.
+     * Return the valid authentication response.
      *
-     * @param  string  $channelName
-     * @param  string  $socketId
-     * @param  array|null  $userData
-     * @return string
+     * @param $request \Illuminate\Http\Request
+     * @param $result mixed
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:162]:
```suggestion
     * @param  \Illuminate\Http\Request  $request
     * @param  mixed  $result
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 5]
```diff
@@ -22,108 +23,161 @@ class AblyBroadcaster extends Broadcaster
...
 
     /**
      * Broadcast the given event.
      *
-     * @param  array  $channels
-     * @param  string  $event
-     * @param  array  $payload
+     * @param $channels array
+     * @param $event string
+     * @param $payload array
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:175]:
```suggestion
     * @param  array  $channels
     * @param  string  $event
     * @param  array  $payload
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 6]
```diff
@@ -139,62 +193,114 @@ public function broadcast(array $channels, $event, array $payload = [])
     }
 
     /**
-     * Build an Ably message object for broadcasting.
+     * @param $channelName string
+     * @param $token string
+     * @param $clientId string
+     * @param $channelCapability string[]
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:199]:
```suggestion
     * @param  string  $channelName
     * @param  string  $token
     * @param  string  $clientId
     * @param  string[]  $channelCapability
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 7]
```diff
@@ -139,62 +193,114 @@ public function broadcast(array $channels, $event, array $payload = [])
...
+        return Utils::generateJwt($header, $claims, $this->getPrivateToken());
+    }
+
+    /**
+     * Remove prefix from channel name.
      *
-     * @param  string  $event
-     * @param  array  $payload
-     * @return \Ably\Models\Message
+     * @param $channel string
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:239]:
```suggestion
     * @param  string  $channel
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 8]
```diff
@@ -139,62 +193,114 @@ public function broadcast(array $channels, $event, array $payload = [])
...
+
+        return $channel;
     }
 
     /**
-     * Return true if the channel is protected by authentication.
+     * Checks if channel is a private channel.
      *
-     * @param  string  $channel
+     * @param $channel string
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:261]:
```suggestion
     * @param  string  $channel
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 9]
```diff
@@ -139,62 +193,114 @@ public function broadcast(array $channels, $event, array $payload = [])
...
+        return Str::startsWith($channel, 'private:');
     }
 
     /**
-     * Remove prefix from channel name.
+     * Checks if channel is a presence channel.
      *
-     * @param  string  $channel
-     * @return string
+     * @param $channel string
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:272]:
```suggestion
     * @param  string  $channel
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 10]
```diff
@@ -139,62 +193,114 @@ public function broadcast(array $channels, $event, array $payload = [])
...
-                        : Str::replaceFirst('presence-', '', $channel);
-        }
+        return Str::startsWith($channel, 'presence:');
+    }
 
-        return $channel;
+    /**
+     * Checks if channel needs authentication.
+     *
+     * @param $channel string
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:283]:
```suggestion
     * @param  string  $channel
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 11]
```diff
@@ -139,62 +193,114 @@ public function broadcast(array $channels, $event, array $payload = [])
...
+    public function isGuardedChannel($channel)
+    {
+        return $this->isPrivateChannel($channel) || $this->isPresenceChannel($channel);
     }
 
     /**
      * Format the channel array into an array of strings.
      *
-     * @param  array  $channels
+     * @param $channels array
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:294]:
```suggestion
     * @param  array  $channels
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 12]
```diff
@@ -204,32 +310,34 @@ protected function formatChannels(array $channels)
     }
 
     /**
-     * Get the public token value from the Ably key.
+     * Build an Ably message object for broadcasting.
      *
-     * @return mixed
+     * @param $event string
+     * @param $payload array
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:316]:
```suggestion
     * @param  string  $event
     * @param  array  $payload
```
First variable type and then the variable name

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php [hunk 13]
```diff
@@ -204,32 +310,34 @@ protected function formatChannels(array $channels)
...
+        });
     }
 
     /**
-     * Get the private token value from the Ably key.
-     *
-     * @return mixed
+     * @param $channelName string
+     * @param $connectionId string
+     * @param $userId string
```
Comment (Jubeki) [inline src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php:331]:
```suggestion
     * @param  string  $channelName
     * @param  string  $connectionId
     * @param  string|null  $userId
```
First variable type and then the variable name
Mark `$userId` as nullable

**Review by Jubeki — COMMENTED** [review]
Please fix some phpdocs.

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 1 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 2 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 3 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 4 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 5 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 6 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 7 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 8 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 9 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 10 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 11 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 12 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

File: src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php (hunk 13 above)
Comment (sacOO7) [reply src/Illuminate/Broadcasting/Broadcasters/AblyBroadcaster.php]: Fixed!

**Comment by taylorotwell** [comment]
So when people update their Ably stuff will stop working unless they opt-in to the Deprecated Ably broadcaster?

**Comment by sacOO7** [comment]
> So when people update their Ably stuff will stop working unless they opt-in to the Deprecated Ably broadcaster?

Yes, it's partially true though. Since you only need to set `ABLY_PUSHER_ADAPTER` to `true`.
If preferred, we could change it so that the `ABLY_PUSHER_ADAPTER` setting is `true` by default for `9.x` and make the breaking change separately as a `10.x` PR.

**Comment by taylorotwell** [comment]
I think we would definitely have to make the old provider the default on a 9.x patch release.

**Comment by sacOO7** [comment]
> I think we would definitely have to make the old provider the default on a 9.x patch release.

Updated with default use of `pusher` based `AblyBroadcaster`. So, no more breaking changes : )

**Comment by taylorotwell** [comment]
One thing that is stopping me from reviewing this PR is the way the existing AblyBroadcaster was modified. It makes it impossible to read through that large diff.

I would suggest making the "new" / dedicated first-party AblyBroadcaster the entirely new file... something like AblyNativeBroadcaster ... that way that file shows as entirely new instead of the "deprecated" broadcaster being the new file 
 and a big mangled diff on the existing broadcaster that is hard to follow and reason about.

**Comment by sacOO7** [comment]
> I would suggest making the "new" / dedicated first-party AblyBroadcaster the entirely new file... something like AblyNativeBroadcaster ... that way that file shows as entirely new instead of the "deprecated" broadcaster being the new file

Sure, we will do it 👍

**Comment by sacOO7** [comment]
> One thing that is stopping me from reviewing this PR is the way the existing AblyBroadcaster was modified. It makes it impossible to read through that large diff.
> 
> I would suggest making the "new" / dedicated first-party AblyBroadcaster the entirely new file... something like AblyNativeBroadcaster ... that way that file shows as entirely new instead of the "deprecated" broadcaster being the new file and a big mangled diff on the existing broadcaster that is hard to follow and reason about.

Updated!

**Comment by sacOO7** [comment]
any updates on the PR @taylorotwell

**Comment by taylorotwell** [comment]
Not at the moment - it's a big PR and I have limited time to review it atm.

**Comment by driesvints** [comment]
@sacOO7 can you maybe rebase with 9.x? Then PHP 8.2 tests can run as well.

**Comment by sacOO7** [comment]
> @sacOO7 can you maybe rebase with 9.x? Then PHP 8.2 tests can run as well.

Sure, we will do it 👍

**Comment by sacOO7** [comment]
> @sacOO7 can you maybe rebase with 9.x? Then PHP 8.2 tests can run as well.

Updated! @driesvints Can you please review the PR? We have put a good amount of work to make this better.

**Comment by sacOO7** [comment]
> > @sacOO7 can you maybe rebase with 9.x? Then PHP 8.2 tests can run as well.
> 
> Sure, we will do it 👍

It seems you need to re-run tests since it was unable to setup [`tests` for php version 8.2](https://github.com/laravel/framework/actions/runs/3072692175/jobs/4964344811).

**Comment by driesvints** [comment]
@sacOO7 done 👍

File: src/Illuminate/Broadcasting/Broadcasters/Ably/Utils.php [hunk 14]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Illuminate\Broadcasting\Broadcasters\Ably;
+
+/**
+ * @author Sachin Shinde (sachinshinde7676@gmail.com)
+ */
```
Comment (driesvints) [inline src/Illuminate/Broadcasting/Broadcasters/Ably/Utils.php:7]: Please remove these docblocks on top of the classes.

File: src/Illuminate/Broadcasting/Broadcasters/Ably/Utils.php [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+class Utils
+{
+    // JWT related PHP utility functions
+    /**
+     * @param  string  $jwt
+     * @return array
+     */
```
Comment (driesvints) [inline src/Illuminate/Broadcasting/Broadcasters/Ably/Utils.php:14]: Please make sure all docblocks are in the same format as the rest of the framework. Each one needs a small one line explanation on top.

File: src/Illuminate/Broadcasting/Broadcasters/Ably/Utils.php [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+        return ['header' => $header, 'payload' => $payload];
+    }
+
+    /**
+     * @param  array  $headers
+     * @param  array  $payload
+     * @return string
+     */
+    public static function generateJwt($headers, $payload, $key)
```
Comment (driesvints) [inline src/Illuminate/Broadcasting/Broadcasters/Ably/Utils.php:29]: If there's parameters that can only be arrays then please add the type hints to the method signature. But do not add other types like string, bool, int, etc.

```suggestion
    public static function generateJwt(array $headers, array $payload, $key)
```

File: src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+    ];
+
+    /**
+     * Create a new broadcaster instance.
+     *
+     * @param  \Ably\AblyRest  $ably
+     * @param  array  $config
+     * @return void
+     */
+    public function __construct(AblyRest $ably, $config)
```
Comment (driesvints) [inline src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php:50]:
```suggestion
    public function __construct(AblyRest $ably, array $config)
```

File: src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+        }
+        if (array_key_exists('disable_public_channels', $config) && $config['disable_public_channels']) {
+            $this->defaultChannelClaims = ['public:*' => ['channel-metadata']];
+        }
+        if (array_key_exists('token_expiry', $config)) {
+            $this->tokenExpiry = $config['token_expiry'];
+        }
+    }
+
+    private static $serverTimeDiff = null;
```
Comment (driesvints) [inline src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php:64]: Move this to the top with the rest of the class properties.

File: src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+        $normalizedChannelName = $this->normalizeChannelName($channelName);
+        $userId = null;
+        $channelCapability = ['*'];
+        $user = $this->retrieveUser($request, $normalizedChannelName);
+        if ($user) {
+            $userId = method_exists($user, 'getAuthIdentifierForBroadcasting')
+                ? $user->getAuthIdentifierForBroadcasting()
+                : $user->getAuthIdentifier();
+        }
+        if ($this->isGuardedChannel($channelName)) {
```
Comment (driesvints) [inline src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php:129]: Always add some breathing room above and below an `if` statement.

```suggestion
        $user = $this->retrieveUser($request, $normalizedChannelName);

        if ($user) {
            $userId = method_exists($user, 'getAuthIdentifierForBroadcasting')
                ? $user->getAuthIdentifierForBroadcasting()
                : $user->getAuthIdentifier();
        }

        if ($this->isGuardedChannel($channelName)) {
```

File: src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+            if (! $user) {
+                throw new AccessDeniedHttpException('User not authenticated, '.$this->stringify($channelName, $connectionId));
+            }
+            try {
+                $userChannelMetaData = parent::verifyUserCanAccessChannel($request, $normalizedChannelName);
+                if (is_array($userChannelMetaData) && array_key_exists('capability', $userChannelMetaData)) {
+                    $channelCapability = $userChannelMetaData['capability'];
+                    unset($userChannelMetaData['capability']);
+                }
+            } catch (\Exception $e) {
```
Comment (driesvints) [inline src/Illuminate/Broadcasting/Broadcasters/AblyNativeBroadcaster.php:139]: Always use FQN:

```suggestion
            } catch (Exception $e) {
```

**Review by driesvints — CHANGES_REQUESTED** [review]
Went through the code style a bit. Taylor will review the functionality.

**Comment by sacOO7** [comment]
Sure, we will update according to mentioned changes @driesvints 👍

**Comment by driesvints** [comment]
Thanks @sacOO7. I'm probably not the best person to review that as my JS skills aren't really that good for it. I also don't know too much about Ably itself.

**Comment by BrandonSurowiec** [comment]
Just a quick tip, you can install Laravel Pint globally `composer global require laravel/pint` and run `pint` in your root directory in the CLI to have most of the formatting issues solved for you.

**Comment by taylorotwell** [comment]
To be honest - something that would make my life much easier is if you could release this as a package that adds this as a custom driver.

All you do is call `Broadcast::extend('ably-native', function () {...})` where the callback returns your broadcaster. I would be willing to help you if you have any questions on how to do that. Then you can publish your broadcaster on Composer and we can mention it in our official documentation.

Then, we just need to find a way for the client-side stuff to also be an NPM package that you can maintain and that plays well with Echo.

**Comment by sacOO7** [comment]
Thanks, @taylorotwell, this really helps! Maybe it's a coincidence, we are trying the same approach. You can check our repository
 https://github.com/ably/laravel-broadcaster/tree/feature/ably-service-provider. 
We will surely need your help once this service provider is ready as a composer package. 
Thanks for the suggestion again!
